### ISSUE TITLE
Double-posted message

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#2374]
Occasionally when someone sends a message from the app, it ends up getting sent twice, creating a duplicate. Here's a recent example on chat.zulip.org:
![image](https://user-images.githubusercontent.com/28173/39069925-46c06ec0-4496-11e8-80b8-e7daf46c4f51.png)
(And the first message has a "13:01" timestamp -- so 16 minutes before the second one.)

**Comment by timabbott** [comment]
Yes, we've confirmed this particular message was sent with the ZulipMobile client.

**Comment by gnprice** [comment]
We had a rather spectacular demonstration of this bug today, provided by @armaanahluwalia :

![image](https://user-images.githubusercontent.com/28173/43671293-fb5347d8-974b-11e8-989a-68e7a3d02f1f.png)

An hour earlier @armaanahluwalia had reported being on 15.0.92 (the current latest prod release) on iOS, so that's presumably the environment where that happened.

**Comment by gnprice** [comment]
We just had another report of this. The user was most likely on the latest version (which went out last week), 20.0.103.

**Comment by timabbott** [comment]
I think that's correct (the server doesn't use it for that purpose) -- I think because on web, we mostly only retransmit after errors after explicit user request.  If we wanted to have the server handle this for us, we could add a database table for tracking such requests and return an "already sent" 200.  

But even with the current server design, I think if the outbox logic only tried to resend things after ensuring it'd processed its event queue, this should be quite rate, since you'd need to have a network interruption longer than 10 minutes to have an issue (and it would become effectively never if we move to the event queues 2.0 system)

**Comment by borisyankov** [comment]
We do track a `isSent` flag on each `Outbox` message which we persist between app restarts.

Even before that we were removing the messages a bit too aggressively from the Outbox, on a successful `POST` even before waiting for the event queue to confirm.

I am curious if the `local_id` of the duplicate messages was the same?

**Comment by gnprice** [comment]
See also #3584, a later report of this with the extra twist that the second post was nearly a week after the original.

**Comment by gnprice** [comment]
Just had a case of this [here](https://chat.zulip.org/#narrow/stream/65-checkins/topic/rohitt/near/827328), with a message I sent. (From v26.22.145 on Android 10.)

I was on a bus in San Francisco -- lots of cell coverage, but lots of devices talking, so it's unsurprising if the connection sometimes fails. As I experienced it, the sequence was:
* I hit send on the message.
* IIRC it takes a few seconds before the spinner disappears.
* When it does, a second copy of the message almost immediately appears too.

As the server saw it, the two copies of the message came 3 seconds apart. (Based on the detailed timestamps shown when I hover over the timestamps in the webapp.)

**Comment by grahamegrieve** [comment]
There's two different things going on here. With the mobile client, there's a thing where it double-posts on unreliable networks, because it doesn't get confirmation. But there's also a UI problem which bit me here: https://chat.fhir.org/#narrow/stream/179219-analytics-on-FHIR/topic/Contained.20resources.20out-of-scope.20for.20the.20initial.20milestone/near/407338972

I was using the mobile app, on a very slow link (airplane).  I typed the message, and hit send. I thought I'd missed, there's no visual queue. So I did it again. Then I realised I wanted to change the message, and then pressed send, and realised... it was already sending. Twice, in fact.

very confusing UI! - and this is very specifically a UI problem unlike the unreliable network - if you send, nothing changes except for a very easy to miss circle icon. And then you can still edit and send. 😢

**Comment by gnprice** [comment]
> hit send. I thought I'd missed, there's no visual queue.

Interesting, thanks for the report.

The behavior I'd expect the app to currently have here is that when you hit send, your message shows up immediately in the message list. There's a small spinning circle at the right of the message to indicate that it hasn't yet finished sending… but other than that, it should look exactly like it will once it does successfully send.

Do I understand correctly that the message didn't show up like that? If so, then that's a bug.

**Comment by grahamegrieve** [comment]
I'm pretty sure it didn't show up like that

**Comment by gnprice** [comment]
OK, thanks for confirming. Sounds like we have a previously-unknown bug where that outgoing message (or "local echo") doesn't always show up.

Because the existing app is [in maintenance mode](https://github.com/zulip/zulip-mobile#maintenance-mode) as we focus on building a new Zulip mobile app to replace it, I think it's unlikely that we'll debug to get to the bottom of that issue. Instead, when we build the corresponding "local echo" feature in the new app, we'll just aim to write it robustly to avoid this sort of issue — and if it develops such an issue anyway, we'll debug and fix it there.

We didn't previously have this "local echo" feature in the new app's tracker. So I've just filed https://github.com/zulip/zulip-flutter/issues/576 for that, and I used the report above as an example of why it's important.
